ThalesGroup/k8s-kms-plugin
# feature : aes-cbc then hmac for tpm2

#### Proposed Changes ####

This pull requests provide a new mechanism for encryption operations :
- AES CBC for encryption / decryption
- HMAC SHA2 for verification

The payloads and authentication tags for the verification are encapsulated in JWE tokens.

#### Types of Changes ####

New feature to support AES CBC then HMAC SHA2.
It enables TPM2 capabilities for the KMS provider of Kubernetes.

#### Verification ####

Using a virtual machine on Debian :

- Install a virtual TPM on you machine
- Create 2 keys in the virtual TPM :
  - One labelled "aes0" as the secret key for encryption
  - One labelled "hmac0" as the hmac key for verification
- Start the k8s-kms-plugin using these labels and the proper arguments :

```sh
k8s-kms-plugin serve \
  --p11-lib $MODULE --p11-label mylabel --p11-pin mypin --p11-key-label aes0 --p11-hmac-label hmac0 \
  --algorithm aes-cbc --enable-server
```

- Start a Kubernetes cluster with a kms provider configuration pointing at the default socket communication with the k8s-kms-plugin in `/tmp/run`. Example with k3s :

```sh
curl -sfL https://get.k3s.io | sh -s - --kube-apiserver-arg=encryption-provider-config=encryption-conf.yaml
```

With the following configuration :

```yaml
apiVersion: apiserver.config.k8s.io/v1
kind: EncryptionConfiguration
resources:
  - resources:
      - secrets
    providers:
      - kms:
          name: kms-server
          endpoint: unix:///tmp/run/hsm-plugin-server.sock
          cachesize: 0
          timeout: 10s
      - identity: {}
```

- Check that the secrets are properly encrypted in the sqlite database of K3S
- Verify the decryption by requesting a secret.

#### Linked Issues ####

#1 

#### User-Facing Change ####

This PR is not breaking change. The default arguments and usage remaines the same.

New arguments are provided to the `server` option of the plugin in order to support the new algorithm and types of keys :

- `--algorithm` : `aes-gcm` or `aes-cbc`. Default is `aes-gcm`.
- `--p11-hmac-label` : key label that provides an hmac key for SHA operations.
